### PR TITLE
Added Optional `sphinx` docs linkcheck

### LINKED ISSUE #8107: Testing URL links in docs & readme
## 🚀 Feature

it may easily happen that some external links are not valid anymore so we shall test them all...

### Motivation

detect broken links in the process

### Alternatives

load the URLs via bs4? (a web scraping tool)
assert that the text you get back contains an expected string?

### PR DESCRIPTION
**Pull request by shenoynikhil**
## What does this PR do?
Fixes #8107 from sphinx pov. Since this sphinx linkcheck can yield false positives, the step added has a `continue-on-error` flag as `True`

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: .github/workflows/docs-checks.yml [hunk 1]
```diff
@@ -158,6 +158,12 @@ jobs:
         run: |
           make html --debug --jobs $(nproc) SPHINXOPTS="-W --keep-going"
 
+      - name: Check External Links in Sphinx Documentation (Optional)
+        run: |
+          cd docs
+          make linkcheck
```
Comment (Borda) [inline .github/workflows/docs-checks.yml:164]:
```suggestion
        working-directory: ./docs/source-${{ matrix.pkg-name }}
        run: make linkcheck
```

Comment (shenoynikhil) [reply .github/workflows/docs-checks.yml]: Updated with this change.

Comment (Borda) [reply .github/workflows/docs-checks.yml]: Btw, you can easily just accept a proposal =)

**Comment by Borda** [comment]
As this workflow is not picked by the PR but using master, can you run it locally and report/fix broken links?

**Comment by shenoynikhil** [comment]
> As this workflow is not picked by the PR but using master, can you run it locally and report/fix broken links?

Ran the following to get broken links,
```bash
cd docs/source-pytorch
make linkcheck >> output.txt
grep -F "broken" output.txt
```
Resulted into this,

```bash
(common/hyperparameters: line   14) broken    ../cli/lightning_cli.html - 
(starter/introduction: line  182) broken    ../common/trainer.html#trainer-flags - 
(common/lightning_module: line 1211) broken    http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.56.7941&rep=rep1&type=pdf - 404 Client Error: Not Found for url: https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.56.7941&rep=rep1&type=pdf
(ecosystem/community_examples: line   13) broken    https://colab.research.google.com/drive/1-_LKx4HwAxl5M6xPJmqAAu444LTDQoa3#scrollTo=BHBz1_AnamN_ - Anchor 'scrollTo=BHBz1_AnamN_' not found
(ecosystem/community_examples: line   19) broken    https://colab.research.google.com/drive/1F_RNcHzTfFuQf-LeKvSlud6x7jXYkG31#scrollTo=yr7eaxkF-djf - Anchor 'scrollTo=yr7eaxkF-djf' not found
(accelerators/ipu_intermediate: line   63) broken    https://docs.graphcore.ai/projects/graph-analyser-userguide/en/latest/graph-analyser.html#opening-reports - 404 Client Error: Not Found for url: https://docs.graphcore.ai/projects/graph-analyser-userguide/en/latest/graph-analyser.html
(accelerators/ipu_advanced: line   29) broken    https://docs.graphcore.ai/projects/poptorch-user-guide/en/latest/reference.html#poptorch.Options.anchorMode - Anchor 'poptorch.Options.anchorMode' not found
(api/pytorch_lightning.strategies.IPUStrategy: line    7) broken    https://docs.graphcore.ai/projects/poptorch-user-guide/en/0.1.67/batching.html - 404 Client Error: Not Found for url: https://docs.graphcore.ai/projects/poptorch-user-guide/en/0.1.67/batching.html
(accelerators/hpu_basic: line   24) broken    https://docs.habana.ai/en/latest/Gaudi_Overview/Gaudi_Overview.html#gaudi-architecture - Anchor 'gaudi-architecture' not found
(accelerators/hpu_basic: line  115) broken    https://docs.habana.ai/en/latest/PyTorch_User_Guide/PyTorch_User_Guide.html#habana-data-loader - Anchor 'habana-data-loader' not found
(api/pytorch_lightning.loggers.neptune: line   57) broken    https://docs.neptune.ai/you-should-know/logging-metadata#essential-logging-methods - Anchor 'essential-logging-methods' not found
(generated/CHANGELOG: line 1329) broken    https://github.com/Lightning-AI/lightning/pull/#10020 - Anchor '10020' not found
(ecosystem/asr_nlp_tts: line  810) broken    https://github.com/NVIDIA/NeMo#tutorials - Anchor 'tutorials' not found
(api/pytorch_lightning.callbacks.ModelPruning: line    3) broken    https://github.com/pytorch/pytorch/blob/1.7.1/torch/nn/utils/prune.py#L1176-L1180 - 404 Client Error: Not Found for url: https://github.com/pytorch/pytorch/blob/1.7.1/torch/nn/utils/prune.py
(api/pytorch_lightning.callbacks.StochasticWeightAveraging: line    1) broken    https://github.com/pytorch/pytorch/blob/v1.7.1/torch/optim/swa_utils.py#L164-L165 - Anchor 'L164-L165' not found
(api/pytorch_lightning.callbacks.StochasticWeightAveraging: line    1) broken    https://github.com/pytorch/pytorch/blob/v1.7.1/torch/optim/swa_utils.py#L104-L112 - Anchor 'L104-L112' not found
(api/pytorch_lightning.callbacks.StochasticWeightAveraging: line    1) broken    https://github.com/pytorch/pytorch/blob/v1.7.1/torch/optim/swa_utils.py#L140-L154 - Anchor 'L140-L154' not found
(api/pytorch_lightning.callbacks.StochasticWeightAveraging: line    1) broken    https://github.com/pytorch/pytorch/blob/v1.7.1/torch/optim/swa_utils.py#L95-L97 - Anchor 'L95-L97' not found
(accelerators/tpu_basic: line  127) broken    https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#get-a-metrics-report - Anchor 'get-a-metrics-report' not found
(accelerators/tpu_basic: line  127) broken    https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#known-performance-caveats - Anchor 'known-performance-caveats' not found
(accelerators/tpu_faq: line  113) broken    https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#troubleshooting - Anchor 'troubleshooting' not found
(accelerators/tpu_advanced: line   53) broken    https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#xla-tensor-quirks - Anchor 'xla-tensor-quirks' not found
(cli/lightning_cli_expert: line  265) broken    https://jsonargparse.readthedocs.io/en/stable/#jsonargparse.ArgumentParser.link_arguments - Anchor 'jsonargparse.ArgumentParser.link_arguments' not found
(ecosystem/asr_nlp_tts: line  447) broken    https://huggingface.co/transformers/model_doc/auto.html#autotokenizer - Anchor 'autotokenizer' not found
(extensions/callbacks: line   64) broken    https://lightning-bolts.readthedocs.io/en/stable/deprecated/callbacks/self_supervised.html#sslonlineevaluator - 404 Client Error: Not Found for url: https://lightning-bolts.readthedocs.io/en/stable/deprecated/callbacks/self_supervised.html
(extensions/callbacks: line   66) broken    https://lightning-bolts.readthedocs.io/en/stable/deprecated/callbacks/variational.html#latent-dim-interpolator - 404 Client Error: Not Found for url: https://lightning-bolts.readthedocs.io/en/stable/deprecated/callbacks/variational.html
(extensions/callbacks: line   65) broken    https://lightning-bolts.readthedocs.io/en/stable/deprecated/callbacks/vision.html#confused-logit - 404 Client Error: Not Found for url: https://lightning-bolts.readthedocs.io/en/stable/deprecated/callbacks/vision.html
(extensions/callbacks: line   67) broken    https://lightning-bolts.readthedocs.io/en/stable/deprecated/callbacks/vision.html#tensorboard-image-generator - 404 Client Error: Not Found for url: https://lightning-bolts.readthedocs.io/en/stable/deprecated/callbacks/vision.html
( ecosystem/flash: line   24) broken    https://lightning-flash.readthedocs.io/en/latest/reference/semantic_segmentation.html#from-folders - Anchor 'from-folders' not found
(clouds/grid_costs: line    4) broken    https://platform.grid.ai/#/dashboard - Anchor '/dashboard' not found
(clouds/run_advanced: line   97) broken    https://platform.grid.ai/#/runs - Anchor '/runs' not found
(ecosystem/transformers: line   19) broken    https://pytorch-lightning.readthedocs.io/en/latest/multi_gpu.html#deepspeed - 404 Client Error: Not Found for url: https://pytorch-lightning.readthedocs.io/en/latest/multi_gpu.html
(ecosystem/transformers: line   19) broken    https://pytorch-lightning.readthedocs.io/en/latest/multi_gpu.html#sharded-training - 404 Client Error: Not Found for url: https://pytorch-lightning.readthedocs.io/en/latest/multi_gpu.html
(common/precision_intermediate: line   60) broken    https://pytorch.org/docs/stable/amp.html#op-specific-behavior - Anchor 'op-specific-behavior' not found
(advanced/pruning_quantization: line   65) broken    https://pytorch.org/docs/stable/quantization.html#quantization-aware-training - Anchor 'quantization-aware-training' not found
(api/pytorch_lightning.callbacks.QuantizationAwareTraining: line   17) broken    https://pytorch.org/docs/stable/torch.quantization.html#torch.quantization.QConfig - 404 Client Error: Not Found for url: https://pytorch.org/docs/stable/torch.quantization.html
(notebooks/lightning_examples/finetuning-scheduler: line 1513) broken    https://tensorboard.dev/experiment/n7U8XhrzRbmvVzC4SQSpWw/#scalars&_smoothingWeight=0&runSelectionState=eyJmdHNfZXhwbGljaXQiOmZhbHNlLCJub2Z0c19iYXNlbGluZSI6ZmFsc2UsImZ0c19pbXBsaWNpdCI6dHJ1ZX0%3D - Anchor 'scalars&_smoothingWeight=0&runSelectionState=eyJmdHNfZXhwbGljaXQiOmZhbHNlLCJub2Z0c19iYXNlbGluZSI6ZmFsc2UsImZ0c19pbXBsaWNpdCI6dHJ1ZX0%3D' not found
(notebooks/lightning_examples/finetuning-scheduler: line 1506) broken    https://tensorboard.dev/experiment/n7U8XhrzRbmvVzC4SQSpWw/#scalars&_smoothingWeight=0&runSelectionState=eyJmdHNfZXhwbGljaXQiOmZhbHNlLCJub2Z0c19iYXNlbGluZSI6dHJ1ZSwiZnRzX2ltcGxpY2l0IjpmYWxzZX0%3D - Anchor 'scalars&_smoothingWeight=0&runSelectionState=eyJmdHNfZXhwbGljaXQiOmZhbHNlLCJub2Z0c19iYXNlbGluZSI6dHJ1ZSwiZnRzX2ltcGxpY2l0IjpmYWxzZX0%3D' not found
(notebooks/lightning_examples/finetuning-scheduler: line 1506) broken    https://tensorboard.dev/experiment/n7U8XhrzRbmvVzC4SQSpWw/#scalars&_smoothingWeight=0&runSelectionState=eyJmdHNfZXhwbGljaXQiOnRydWUsIm5vZnRzX2Jhc2VsaW5lIjpmYWxzZSwiZnRzX2ltcGxpY2l0IjpmYWxzZX0%3D - Anchor 'scalars&_smoothingWeight=0&runSelectionState=eyJmdHNfZXhwbGljaXQiOnRydWUsIm5vZnRzX2Jhc2VsaW5lIjpmYWxzZSwiZnRzX2ltcGxpY2l0IjpmYWxzZX0%3D' not found
(advanced/model_parallel: line  616) broken    https://www.deepspeed.ai/news/2021/03/07/zero3-offload.html - 404 Client Error: Not Found for url: https://www.deepspeed.ai/news/2021/03/07/zero3-offload.html
(notebooks/lightning_examples/augmentation_kornia: line  128) broken    https://www.kornia.org - HTTPSConnectionPool(host='www.kornia.org', port=443): Max retries exceeded with url: / (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1131)')))
(starter/introduction: line  102) broken    installation.html - 
(           index: line   65) broken    starter/installation.html - 
```

**Comment by shenoynikhil** [comment]
I think a separate issue would be required to do a cleanup of these. What do you think?

**Comment by shenoynikhil** [comment]
With `linkcheck_anchors=False` in conf.py (Based on [this](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-linkcheck_anchors) and this [issue](https://github.com/sphinx-doc/sphinx/issues/8268).) 

Only the following links are broken.
```
(common/hyperparameters: line   14) broken    ../cli/lightning_cli.html - 
(starter/introduction: line  182) broken    ../common/trainer.html#trainer-flags - 
(common/lightning_module: line 1211) broken    http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.56.7941&rep=rep1&type=pdf - 404 Client Error: Not Found for url: https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.56.7941&rep=rep1&type=pdf
(common/hyperparameters: line   14) broken    ../cli/lightning_cli.html - 
(starter/introduction: line  182) broken    ../common/trainer.html#trainer-flags - 
(common/lightning_module: line 1211) broken    http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.56.7941&rep=rep1&type=pdf - 404 Client Error: Not Found for url: https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.56.7941&rep=rep1&type=pdf
(accelerators/ipu_intermediate: line   63) broken    https://docs.graphcore.ai/projects/graph-analyser-userguide/en/latest/graph-analyser.html#opening-reports - 404 Client Error: Not Found for url: https://docs.graphcore.ai/projects/graph-analyser-userguide/en/latest/graph-analyser.html
(api/pytorch_lightning.strategies.IPUStrategy: line    7) broken    https://docs.graphcore.ai/projects/poptorch-user-guide/en/0.1.67/batching.html - 404 Client Error: Not Found for url: https://docs.graphcore.ai/projects/poptorch-user-guide/en/0.1.67/batching.html
(api/pytorch_lightning.callbacks.ModelPruning: line    3) broken    https://github.com/pytorch/pytorch/blob/1.7.1/torch/nn/utils/prune.py#L1176-L1180 - 404 Client Error: Not Found for url: https://github.com/pytorch/pytorch/blob/1.7.1/torch/nn/utils/prune.py
(extensions/callbacks: line   65) broken    https://lightning-bolts.readthedocs.io/en/stable/deprecated/callbacks/vision.html#confused-logit - 404 Client Error: Not Found for url: https://lightning-bolts.readthedocs.io/en/stable/deprecated/callbacks/vision.html
(extensions/callbacks: line   67) broken    https://lightning-bolts.readthedocs.io/en/stable/deprecated/callbacks/vision.html#tensorboard-image-generator - 404 Client Error: Not Found for url: https://lightning-bolts.readthedocs.io/en/stable/deprecated/callbacks/vision.html
(extensions/callbacks: line   66) broken    https://lightning-bolts.readthedocs.io/en/stable/deprecated/callbacks/variational.html#latent-dim-interpolator - 404 Client Error: Not Found for url: https://lightning-bolts.readthedocs.io/en/stable/deprecated/callbacks/variational.html
(extensions/callbacks: line   64) broken    https://lightning-bolts.readthedocs.io/en/stable/deprecated/callbacks/self_supervised.html#sslonlineevaluator - 404 Client Error: Not Found for url: https://lightning-bolts.readthedocs.io/en/stable/deprecated/callbacks/self_supervised.html
(ecosystem/transformers: line   19) broken    https://pytorch-lightning.readthedocs.io/en/latest/multi_gpu.html#sharded-training - 404 Client Error: Not Found for url: https://pytorch-lightning.readthedocs.io/en/latest/multi_gpu.html
(ecosystem/transformers: line   19) broken    https://pytorch-lightning.readthedocs.io/en/latest/multi_gpu.html#deepspeed - 404 Client Error: Not Found for url: https://pytorch-lightning.readthedocs.io/en/latest/multi_gpu.html
(api/pytorch_lightning.callbacks.QuantizationAwareTraining: line   17) broken    https://pytorch.org/docs/stable/torch.quantization.html#torch.quantization.QConfig - 404 Client Error: Not Found for url: https://pytorch.org/docs/stable/torch.quantization.html
(advanced/model_parallel: line  616) broken    https://www.deepspeed.ai/news/2021/03/07/zero3-offload.html - 404 Client Error: Not Found for url: https://www.deepspeed.ai/news/2021/03/07/zero3-offload.html
(notebooks/lightning_examples/augmentation_kornia: line  128) broken    https://www.kornia.org - HTTPSConnectionPool(host='www.kornia.org', port=443): Max retries exceeded with url: / (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1131)')))
(starter/introduction: line  102) broken    installation.html - 
(           index: line   65) broken    starter/installation.html - 
```

**Comment by awaelchli** [comment]
> I think a separate issue would be required to do a cleanup of these. Or should I just do the cleanup here?

I think either way is fine

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Pull request merged**